### PR TITLE
DEV: Refactor header offset calculations

### PR DESCRIPTION
**Pull request by pmusaraj**
This introduces a new CSS var, `--main-outlet-offset` alongside the existing `--header-offset`. 

- `--header-offset` remains as before, giving in CSS the header's offset from the top of the viewpoint
- `--main-outlet-offset` includes `--header-offset` plus any elements that are above the main outlet and below the header, for example, elements in the `below-header` outlet

This fixes issues with the sidebar overflowing on sites with some theme components. And incompatibilities between said theme components and chat. 

Before/After (notice the missing bottom of chat and sidebar in the before screenshot)

image

image

### CODE REVIEW

File: app/assets/javascripts/discourse/app/components/glimmer-site-header.gjs [hunk 1]
```diff
@@ -119,33 +123,24 @@ export default class GlimmerSiteHeader extends Component {
...
     this._headerWrap = document.querySelector(".d-header-wrap");
+    this._mainOutletWrapper = document.querySelector("#main-outlet-wrapper");
     if (this._headerWrap) {
-      schedule("afterRender", () => {
-        this.headerElement = this._headerWrap.querySelector("header.d-header");
-        this.updateHeaderOffset();
-        document.documentElement.style.setProperty(
-          "--header-top",
-          `${this.headerElement.offsetTop}px`
-        );
```
Comment (pmusaraj) [inline app/assets/javascripts/discourse/app/components/glimmer-site-header.gjs:129]: This is no longer useful. It was used in the menu panel only, but a) it always returns 0, b) was supposed to apply to mobile menu panels but wasn't used in the mobile stylesheets.

File: app/assets/javascripts/discourse/app/components/glimmer-site-header.gjs [hunk 2]
```diff
@@ -119,33 +123,24 @@ export default class GlimmerSiteHeader extends Component {
...
-      });
+      this._recalculateHeaderOffset();
 
-      window.addEventListener("scroll", this._onScroll, {
+      window.addEventListener("scroll", this._recalculateHeaderOffset, {
         passive: true,
       });
 
+      this.appEvents.on("page:changed", this, this._recalculateHeaderOffset);
```
Comment (pmusaraj) [inline app/assets/javascripts/discourse/app/components/glimmer-site-header.gjs:134]: Above/below header elements can be added/removed depending on the route, it's good to recalc.

File: app/assets/stylesheets/common/base/menu-panel.scss [hunk 3]
```diff
@@ -16,7 +16,7 @@
   // positions are relative to the .d-header .panel div
   top: 100%; // directly underneath .panel
   right: -10px; // 10px to the right of .panel - adjust as needed
-  max-height: 80vh;
+  max-height: calc(100vh - var(--header-offset) - 1em);
```
Comment (pmusaraj) [inline app/assets/stylesheets/common/base/menu-panel.scss:19]: A slightly unrelated change. `80vh` here was overriding the header calculations, and in some cases cutting off the menu quite early. 

Before/after

image

image

Comment (davidtaylorhq) [reply app/assets/stylesheets/common/base/menu-panel.scss:19]: Should we use dvh here, to ensure the menu won't overflow on devices with dynamically showing/hiding browser chrome? (i.e. iOS)

File: app/assets/javascripts/discourse/app/components/glimmer-site-header.gjs (hunk 2 above)
Comment (davidtaylorhq) [reply app/assets/javascripts/discourse/app/components/glimmer-site-header.gjs:138]: They can also be added/removed in other situations... but it's hard to know how we should handle that efficiently...

Could/should we wrap an element around all the PluginOutlets, and then use a ResizeObserver? 🤔

File: app/assets/stylesheets/common/base/menu-panel.scss (hunk 3 above)
Comment (pmusaraj) [reply app/assets/stylesheets/common/base/menu-panel.scss:19]: Sure, I don't see a downside to using `dvh`.

File: app/assets/javascripts/discourse/app/components/glimmer-site-header.gjs (hunk 2 above)
Comment (pmusaraj) [reply app/assets/javascripts/discourse/app/components/glimmer-site-header.gjs:138]: Use a ResizeObserver on all PluginOutlets? I'm not sure, sounds like a lot of observers.

Comment (davidtaylorhq) [reply app/assets/javascripts/discourse/app/components/glimmer-site-header.gjs:138]: No, I mean add a single wrapper element around everything which sits above the main-outlet. Then add a single ResizeObserver to that.

(but... maybe that wrapper element would cause layout issues... I dunno)

Comment (pmusaraj) [reply app/assets/javascripts/discourse/app/components/glimmer-site-header.gjs:138]: Ah I see. Maybe we could add the observer to `.discourse-root`? Its height _should_ change when elements are added/removed, or when the element's height changes, anyway.

Comment (davidtaylorhq) [reply app/assets/javascripts/discourse/app/components/glimmer-site-header.gjs:138]: Oh yes, good idea!

Comment (pmusaraj) [reply app/assets/javascripts/discourse/app/components/glimmer-site-header.gjs]: Ok, I added a refactor @davidtaylorhq. We had a resize observer on the header wrap element, I remove that and replaced it with one for `.discourse-root`. It should cover all cases. 

Note that here we also have a scroll listener. That one fires a lot more often than the resize observer, and we didn't have a debouncer for it, so I added one. 

Also noticed one more thing. We have a second scroll listener on this file to add a `docked` class to the body. I don't think we are using the `docked` class on the body anymore, and in fact the scroll listener doesn't seem to be working properly. Will address this in a separate PR though.

**Review by davidtaylorhq — APPROVED** [review]
